pkg/ttl/cache/table.go

```diff
@@ -185,28 +185,98 @@ func (t *PhysicalTable) ValidateKeyPrefix(key []types.Datum) error {
 	return nil
 }
 
-// EvalExpireTime returns the expired time
+var mockExpireTimeKey struct{}
+
+// SetMockExpireTime can only used in test
+func SetMockExpireTime(ctx context.Context, tm time.Time) context.Context {
+	return context.WithValue(ctx, mockExpireTimeKey, tm)
+}
+
+// EvalExpireTime returns the expired time.
+// It uses the global timezone to compute the expired time.
+// Then we'll reset the returned expired time to the same timezone as the input `now`.
 func (t *PhysicalTable) EvalExpireTime(ctx context.Context, se session.Session,
 	now time.Time) (expire time.Time, err error) {
-	tz := se.GetSessionVars().Location()
+	if intest.InTest {
+		if tm, ok := ctx.Value(mockExpireTimeKey).(time.Time); ok {
+			return tm, err
+		}
+	}
 
 	expireExpr := t.TTLInfo.IntervalExprStr
 	unit := ast.TimeUnitType(t.TTLInfo.IntervalTimeUnit)
 
+	// Use the global time zone to compute expire time.
+	// Different timezones may have different results event with the same "now" time and TTL expression.
+	// Consider a TTL setting with the expiration `INTERVAL 1 MONTH`.
+	// If the current timezone is `Asia/Shanghai` and now is `2021-03-01 00:00:00 +0800`
+	// the expired time should be `2021-02-01 00:00:00 +0800`, corresponding to UTC time `2021-01-31 16:00:00 UTC`.
+	// But if we use the `UTC` time zone, the current time is `2021-02-28 16:00:00 UTC`,
+	// and the expired time should be `2021-01-28 16:00:00 UTC` that is not the same the previous one.
+	globalTz, err := se.GlobalTimeZone(ctx)
+	if err != nil {
+		return time.Time{}, err
+	}
+
 	var rows []chunk.Row
+
+	// We should set the session time zone to UTC because the next SQLs should be executed in the UTC timezone.
+	// The session time zone should be reverted to the original one after the SQLs are executed.
+	rows, err = se.ExecuteSQL(ctx, "SELECT @@time_zone")
+	if err != nil {
+		return
+	}
+
+	originalTZ := rows[0].GetString(0)
+	if _, err = se.ExecuteSQL(ctx, "SET @@time_zone='UTC'"); err != nil {
+		return
+	}
+
+	defer func() {
+		_, restoreErr := se.ExecuteSQL(ctx, "SET @@time_zone=%?", originalTZ)
+		if err == nil {
+			err = restoreErr
+		}
+	}()
+
+	// Firstly, we should use the UTC time zone to compute the expired time to avoid time shift caused by DST.
+	// The start time should be a time with the same datetime string as `now` but it is in the UTC timezone.
+	nowInGlobalTZ := now.In(globalTz)
+	startTime := time.Date(
```

Comment:
I don't really get the point here: `nowInGlobalTZ` and `startTime` are `time.Time`, what's the difference between this and `startTime := time.In(UTC)`?

Reply:
For example, if global time zone is `Asia/Shanghai` and nowInGlobalTZ is `2020-01-01 08:00:00 +08:00`.

`nowInGlobalTZ.In(time.UTC)` is: `2020-01-01 00:00:00 +00:00`
`startTime` is: `2020-01-01 08:00:00 +00:00`

`startTime` should have a same formated time string with `nowInGlobalTZ` but it they are not the same point

Reply:
This is a bit confusing, could you please add your example to the comment?

Reply:
Done